File: update.js
/**
 * update.js
 * 全台データをHTTP取得し history.json に蓄積、graph.html を更新する
 */
const https = require('https');
const fs = require('fs');
const path = require('path');

const DIR = __dirname;
const UNIT_LIST_URL = 'https://daidata.goraggio.com/100563/unit_list?model=e%E7%89%99%E7%8B%BC12%20XX-MJ&ballPrice=4.00&ps=P';
const UNITS_FILE = path.join(DIR, 'units.json');
const UNITS_FALLBACK = [41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57];
const HISTORY_FILE = path.join(DIR, 'history.json');
const TEMPLATE_FILE = path.join(DIR, 'graph_template.html');
const OUTPUT_FILE = path.join(DIR, 'graph.html');
const PRED_HISTORY_FILE = path.join(DIR, 'prediction_history.json');
const PRED_TEMPLATE_FILE = path.join(DIR, 'predictions_template.html');
const PRED_OUTPUT_FILE = path.join(DIR, 'predictions.html');
const KEEP_DAYS = 40; // 1ヶ月＋余裕

// 予測スコアのデフォルト重み
const DEFAULT_WEIGHTS = { consecutiveLoss: 20, recentCumLoss: 1.0, avgDaily: 1.0 };

// ============================================================
// HTTP取得
// ============================================================
function fetchHtml(url) {
  return new Promise((resolve, reject) => {
    const req = https.get(url, {
      headers: {
        'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 Chrome/120.0.0.0 Safari/537.36',
        'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
        'Accept-Language': 'ja,en-US;q=0.7,en;q=0.3',
        'Referer': 'https://daidata.goraggio.com/',
      },
    }, (res) => {
      if (res.statusCode === 301 || res.statusCode === 302) {
        return fetchHtml(res.headers.location).then(resolve).catch(reject);
      }
      const chunks = [];
      res.on('data', c => chunks.push(c));
      res.on('end', () => resolve(Buffer.concat(chunks).toString('utf8')));
    });
    req.on('error', reject);
    req.setTimeout(30000, () => { req.destroy(); reject(new Error('timeout')); });
  });
}

// ============================================================
// HTMLからjqplotデータを抽出（全日別グラフを取得）
// ============================================================
function parseHtml(html) {
  const days = {};

  // 全jqplot呼び出しをマッチ
  const matches = [...html.matchAll(/\.jqplot\s*\(\s*(\[\[[\s\S]*?\]\])\s*,/g)];
  for (const m of matches) {
    let rawData;
    try { rawData = JSON.parse(m[1]); } catch (e) { continue; }
    if (!rawData.length) continue;

    const series = Array.isArray(rawData[0][0]) ? rawData[0] : rawData;
    if (!series.length) continue;

    if (typeof series[0][0] === 'string') {
      // 日別グラフ: [["YYYY-MM-DD HH:MM:SS", value], ...]
      for (const [ts, value] of series) {
        const date = ts.slice(0, 10);
        if (!days[date]) days[date] = [];
        days[date].push(value);
      }
    }
    // 週間グラフ（index形式）はスキップ（日別グラフで補完済み）
  }

  if (Object.keys(days).length === 0) return null;
  return { days };
}

// ============================================================
// "M/D" を "YYYY-MM-DD" に変換
// ============================================================
function normalizeDate(mdStr, fetchDate) {
  const [m, d] = mdStr.split('/').map(Number);
  const year = fetchDate.getFullYear();
  const candidate = new Date(year, m - 1, d);
  // 取得日より7日以上先なら前年
  if (candidate.getTime() > fetchDate.getTime() + 7 * 86400000) {
    return `${year - 1}-${String(m).padStart(2, '0')}-${String(d).padStart(2, '0')}`;
  }
  return `${year}-${String(m).padStart(2, '0')}-${String(d).padStart(2, '0')}`;
}

// ============================================================
// 台リストを動的取得
// ============================================================
async function fetchUnitList() {
  const html = await fetchHtml(UNIT_LIST_URL);
  const matches = [...html.matchAll(/detail\?unit=(\d+)/g)];
  const units = [...new Set(matches.map(m => parseInt(m[1])))].sort((a, b) => a - b);
  if (units.length === 0) throw new Error('台リストが空です');
  return units;
}

// ============================================================
// 全台取得
// ============================================================
async function fetchAll(fetchDate, units) {
  const results = [];
  for (const unit of units) {
    const url = `https://daidata.goraggio.com/100563/detail?unit=${unit}`;
    process.stdout.write(`取得中: 台${unit} ... `);
    try {
      const html = await fetchHtml(url);
const parsed = parseHtml(html);
      if (parsed) {
        results.push({ unit, days: parsed.days });
        const totalPts = Object.values(parsed.days).reduce((s, a) => s + a.length, 0);
        process.stdout.write(`OK (${totalPts}ポイント)\n`);
      } else {
        process.stdout.write('データなし\n');
      }
    } catch (e) {
      process.stdout.write(`エラー: ${e.message}\n`);
    }
    await new Promise(r => setTimeout(r, 500));
  }
  return results;
}

// ============================================================
// history.json にマージ（古いデータを削除）
// ============================================================
function mergeHistory(existing, newData, fetchDate) {
  // unit→{days} のマップに変換
  const map = {};
  for (const u of (existing || [])) map[u.unit] = { ...u };

  for (const u of newData) {
    if (!map[u.unit]) map[u.unit] = { unit: u.unit, days: {} };
    Object.assign(map[u.unit].days, u.days);
  }

  // KEEP_DAYS 日より古いデータを削除
  const cutoff = new Date(fetchDate.getTime() - KEEP_DAYS * 86400000)
    .toISOString().slice(0, 10);
  for (const u of Object.values(map)) {
    for (const date of Object.keys(u.days)) {
      if (date < cutoff) delete u.days[date];
    }
  }

  return Object.values(map).sort((a, b) => a.unit - b.unit);
}

// ============================================================
// graph.html を更新
// ============================================================
function updateHtml(history) {
  const template = fs.readFileSync(TEMPLATE_FILE, 'utf8');
  const html = template.replace('__GRAPH_DATA__', JSON.stringify(history));
  fs.writeFileSync(OUTPUT_FILE, html);
  console.log('graph.html を更新しました');
}

// ============================================================
// JST日付文字列を取得 (YYYY-MM-DD)
// ============================================================
function getJSTDateStr(date) {
  return new Date(date).toLocaleString('sv', { timeZone: 'Asia/Tokyo' }).slice(0, 10);
}

// ============================================================
// 予測スコア計算（サーバーサイド）
// ============================================================
function calcPrediction(history, weights = DEFAULT_WEIGHTS) {
  const allDates = [...new Set(history.flatMap(u => Object.keys(u.days)))].sort();
  const recentDates = allDates.slice(-7);

  return history.map(u => {
    const dayResults = allDates.map(date => {
      const pts = u.days[date];
      return pts && pts.length > 0 ? pts[pts.length - 1] : null;
    }).filter(v => v !== null);

    if (dayResults.length === 0) return null;

    const recentResults = recentDates.map(date => {
      const pts = u.days[date];
      return pts && pts.length > 0 ? pts[pts.length - 1] : null;
    }).filter(v => v !== null);

    let consecutiveLoss = 0;
    for (let i = recentDates.length - 1; i >= 0; i--) {
      const pts = u.days[recentDates[i]];
      const v = pts && pts.length > 0 ? pts[pts.length - 1] : null;
      if (v === null) break;
      if (v < 0) consecutiveLoss++; else break;
    }

    const recentCumLoss = recentResults.reduce((s, v) => s + Math.min(v, 0), 0);
    const avgDaily = dayResults.reduce((s, v) => s + v, 0) / dayResults.length;
    const recentAvg = recentResults.length > 0
      ? recentResults.reduce((s, v) => s + v, 0) / recentResults.length
      : avgDaily;
    const wins  = recentResults.filter(v => v > 0).length;
    const losses = recentResults.filter(v => v < 0).length;

    const score =
      consecutiveLoss * weights.consecutiveLoss
      + Math.max(0, -recentCumLoss) / 1000 * weights.recentCumLoss
      + Math.max(0, -avgDaily) / 500 * weights.avgDaily;

    return {
      unit: u.unit,
      score: Math.round(score * 10) / 10,
      consecutiveLoss,
      recentAvg: Math.round(recentAvg),
      recentCumLoss: Math.round(recentCumLoss),
      wins,
      losses,
    };
  }).filter(Boolean).sort((a, b) => b.score - a.score);
}

// ============================================================
// 予測結果を記録（当日の実績を過去の予測エントリに記入）
// ============================================================
function fillResults(predHistory, history, todayJST) {
  for (const entry of predHistory) {
    if (entry.predDate === todayJST && entry.results === null) {
      entry.results = {};
      for (const u of history) {
        const pts = u.days[todayJST];
        if (pts && pts.length > 0) {
          entry.results[String(u.unit)] = pts[pts.length - 1];
        }
      }
      const filled = Object.keys(entry.results).length;
      console.log(`予測結果記録: ${todayJST} (${filled}台)`);
    }
  }
}

// ============================================================
// PDCA: 過去の正解率からスコア重みを調整
// ============================================================
function calcPdcaWeights(predHistory) {
  const completed = predHistory.filter(
    e => e.results && Object.keys(e.results).length > 0
  ).slice(-14);

  if (completed.length < 5) return DEFAULT_WEIGHTS;

  // 各ファクターのホット台ヒット率を計算
  const factorHits = { consecutiveLoss: { hit: 0, total: 0 }, recentCumLoss: { hit: 0, total: 0 }, avgDaily: { hit: 0, total: 0 } };

  for (const entry of completed) {
    const hotPreds = entry.predictions.slice(0, 3);
    for (const p of hotPreds) {
      const result = entry.results[String(p.unit)];
      if (result === undefined) continue;
      const isHit = result > 0;
      for (const factor of Object.keys(factorHits)) {
        if (p[factor] !== undefined) {
          factorHits[factor].total++;
          if (isHit) factorHits[factor].hit++;
        }
      }
    }
  }

  // 全体ヒット率
  const totalHits  = Object.values(factorHits).reduce((s, f) => s + f.hit, 0);
  const totalTotal = Object.values(factorHits).reduce((s, f) => s + f.total, 0);
  const baseRate   = totalTotal > 0 ? totalHits / totalTotal : 0.5;

  const newWeights = { ...DEFAULT_WEIGHTS };
  for (const [factor, stat] of Object.entries(factorHits)) {
    if (stat.total === 0) continue;
    const hitRate = stat.hit / stat.total;
    // ヒット率が高いほど重みを上げる（最大2倍、最小0.5倍）
    const multiplier = Math.max(0.5, Math.min(2.0, 0.5 + hitRate / Math.max(baseRate, 0.1)));
    newWeights[factor] = Math.round(DEFAULT_WEIGHTS[factor] * multiplier * 10) / 10;
  }

  return newWeights;
}

// ============================================================
// predictions.html を更新
// ============================================================
function updatePredictionsHtml(predHistory) {
  if (!fs.existsSync(PRED_TEMPLATE_FILE)) return;
  const template = fs.readFileSync(PRED_TEMPLATE_FILE, 'utf8');
  const html = template.replace('__PREDICTION_DATA__', JSON.stringify(predHistory));
  fs.writeFileSync(PRED_OUTPUT_FILE, html);
  console.log('predictions.html を更新しました');
}

// ============================================================
// メイン
// ============================================================
(async () => {
  const fetchDate = new Date();
  const now = fetchDate.toLocaleString('ja-JP', { timeZone: 'Asia/Tokyo' });
  console.log(`[${now}] データ取得開始`);

  // 台リストを動的取得
  let activeUnits;
  try {
    activeUnits = await fetchUnitList();
    console.log(`台リスト取得: ${activeUnits.length}台 [${activeUnits.join(', ')}]`);
    const prev = fs.existsSync(UNITS_FILE)
      ? JSON.parse(fs.readFileSync(UNITS_FILE, 'utf8'))
      : activeUnits;
    const added   = activeUnits.filter(u => !prev.includes(u));
    const removed = prev.filter(u => !activeUnits.includes(u));
    if (added.length)   console.log(`台追加: ${added.join(', ')}`);
    if (removed.length) console.log(`台削除: ${removed.join(', ')}`);
    fs.writeFileSync(UNITS_FILE, JSON.stringify(activeUnits));
  } catch (e) {
    console.log(`警告: 台リスト取得失敗 (${e.message})。前回リストを使用。`);
    activeUnits = fs.existsSync(UNITS_FILE)
      ? JSON.parse(fs.readFileSync(UNITS_FILE, 'utf8'))
      : UNITS_FALLBACK;
  }

  const newData = await fetchAll(fetchDate, activeUnits);
  console.log(`\n取得完了: ${newData.length}台`);

  const existing = fs.existsSync(HISTORY_FILE)
    ? JSON.parse(fs.readFileSync(HISTORY_FILE, 'utf8'))
    : [];

  if (newData.length === 0) {
    console.log('警告: 新規データなし。既存の history.json でグラフを更新します。');
    if (existing.length === 0) {
      console.log('エラー: 既存データもありません。終了します。');
      process.exit(1);
    }
    updateHtml(existing);
    updatePredictionsHtml(
      fs.existsSync(PRED_HISTORY_FILE)
        ? JSON.parse(fs.readFileSync(PRED_HISTORY_FILE, 'utf8'))
        : []
    );
    console.log('完了（既存データ使用）');
    return;
  }

  const history = mergeHistory(existing, newData, fetchDate);
  fs.writeFileSync(HISTORY_FILE, JSON.stringify(history));
  console.log(`history.json を更新しました（${activeUnits.length}台 × 最大${KEEP_DAYS}日）`);

  updateHtml(history);

  // ============================================================
  // 予測履歴の更新
  // ============================================================
  const todayJST = getJSTDateStr(fetchDate);
  const tomorrowJST = getJSTDateStr(new Date(fetchDate.getTime() + 86400000));

  const predHistory = fs.existsSync(PRED_HISTORY_FILE)
    ? JSON.parse(fs.readFileSync(PRED_HISTORY_FILE, 'utf8'))
    : [];

  // 当日の実績を過去の予測エントリに記入
  fillResults(predHistory, history, todayJST);

  // PDCA: 過去実績から重みを調整
  const weights = calcPdcaWeights(predHistory);
  console.log(`PDCA重み: consecutiveLoss=${weights.consecutiveLoss}, recentCumLoss=${weights.recentCumLoss}, avgDaily=${weights.avgDaily}`);

  // 翌日の予測エントリを追加（重複しない場合のみ）
  if (!predHistory.find(e => e.predDate === tomorrowJST)) {
    const predictions = calcPrediction(history, weights).slice(0, 9);
    predHistory.push({ predDate: tomorrowJST, madeAt: todayJST, weights, predictions, results: null });
    console.log(`翌日予測を作成: ${tomorrowJST} (上位${predictions.length}台)`);
  }

  // 90日より古いエントリを削除
  const predCutoff = getJSTDateStr(new Date(fetchDate.getTime() - 90 * 86400000));
  const prunedPredHistory = predHistory.filter(e => e.predDate >= predCutoff);

  fs.writeFileSync(PRED_HISTORY_FILE, JSON.stringify(prunedPredHistory));
  console.log('prediction_history.json を更新しました');

  updatePredictionsHtml(prunedPredHistory);
  console.log('完了');
})();

File: update.py
#!/usr/bin/env python3
"""
update.py
全台データをHTTP取得し history.json に蓄積、graph.html を更新する
"""
import json
import os
import re
import sys
import time
import urllib.request
import urllib.error
from datetime import datetime, timezone, timedelta

DIR = os.path.dirname(os.path.abspath(__file__))
UNIT_LIST_URL = 'https://daidata.goraggio.com/100563/unit_list?model=e%E7%89%99%E7%8B%BC12%20XX-MJ&ballPrice=4.00&ps=P'
UNITS_FILE = os.path.join(DIR, 'units.json')
UNITS_FALLBACK = [11,12,13,14,15,16,17,18,19,20,52,53,54,55,56,57,58,59,60]
HISTORY_FILE = os.path.join(DIR, 'history.json')
TEMPLATE_FILE = os.path.join(DIR, 'graph_template.html')
OUTPUT_FILE = os.path.join(DIR, 'graph.html')
PRED_HISTORY_FILE = os.path.join(DIR, 'prediction_history.json')
PRED_TEMPLATE_FILE = os.path.join(DIR, 'predictions_template.html')
PRED_OUTPUT_FILE = os.path.join(DIR, 'predictions.html')
KEEP_DAYS = 40

DEFAULT_WEIGHTS = {'consecutiveLoss': 20, 'recentCumLoss': 1.0, 'avgDaily': 1.0}
JST = timezone(timedelta(hours=9))
HEADERS = {
    'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 Chrome/120.0.0.0 Safari/537.36',
    'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
    'Accept-Language': 'ja,en-US;q=0.7,en;q=0.3',
    'Referer': 'https://daidata.goraggio.com/',
}

# ============================================================
# HTTP取得
# ============================================================
def fetch_html(url):
    req = urllib.request.Request(url, headers=HEADERS)
    try:
        with urllib.request.urlopen(req, timeout=30) as res:
            return res.read().decode('utf-8')
    except urllib.error.URLError as e:
        raise Exception(str(e))

# ============================================================
# 台リストを動的取得
# ============================================================
def fetch_unit_list():
    html = fetch_html(UNIT_LIST_URL)
    matches = re.findall(r'detail\?unit=(\d+)', html)
    units = sorted(set(int(m) for m in matches))
    if not units:
        raise Exception('台リストが空です')
    return units

# ============================================================
# HTMLからjqplotデータを抽出
# ============================================================
def parse_html(html):
    days = {}
    matches = re.findall(r'\.jqplot\s*\(\s*(\[\[[\s\S]*?\]\])\s*,', html)
    for m in matches:
        try:
            raw_data = json.loads(m)
        except Exception:
            continue
        if not raw_data:
            continue
        series = raw_data[0] if isinstance(raw_data[0][0], list) else raw_data
        if not series:
            continue
        if isinstance(series[0][0], str):
            for ts, value in series:
                date = ts[:10]
                if date not in days:
                    days[date] = []
                days[date].append(value)
    if not days:
        return None
    return {'days': days}

# ============================================================
# JST日付文字列を取得
# ============================================================
def get_jst_date_str(dt=None):
    if dt is None:
        dt = datetime.now(JST)
    return dt.strftime('%Y-%m-%d')

# ============================================================
# 全台取得
# ============================================================
def fetch_all(units):
    results = []
    for unit in units:
        url = f'https://daidata.goraggio.com/100563/detail?unit={unit}'
        sys.stdout.write(f'取得中: 台{unit} ... ')
        sys.stdout.flush()
        try:
            html = fetch_html(url)
            parsed = parse_html(html)
            if parsed:
                results.append({'unit': unit, 'days': parsed['days']})
                total_pts = sum(len(v) for v in parsed['days'].values())
                print(f'OK ({total_pts}ポイント)')
            else:
                print('データなし')
        except Exception as e:
            print(f'エラー: {e}')
        time.sleep(0.5)
    return results

# ============================================================
# history.json にマージ
# ============================================================
def merge_history(existing, new_data, fetch_date):
    map_ = {}
    for u in (existing or []):
        map_[u['unit']] = {'unit': u['unit'], 'days': dict(u['days'])}
    for u in new_data:
        if u['unit'] not in map_:
            map_[u['unit']] = {'unit': u['unit'], 'days': {}}
        map_[u['unit']]['days'].update(u['days'])

    cutoff = (fetch_date - timedelta(days=KEEP_DAYS)).strftime('%Y-%m-%d')
    for u in map_.values():
        u['days'] = {d: v for d, v in u['days'].items() if d >= cutoff}

    return sorted(map_.values(), key=lambda u: u['unit'])

# ============================================================
# graph.html を更新
# ============================================================
def update_html(history):
    with open(TEMPLATE_FILE, 'r', encoding='utf-8') as f:
        template = f.read()
    html = template.replace('__GRAPH_DATA__', json.dumps(history))
    with open(OUTPUT_FILE, 'w', encoding='utf-8') as f:
        f.write(html)
    print('graph.html を更新しました')

# ============================================================
# predictions.html を更新
# ============================================================
def update_predictions_html(pred_history):
    if not os.path.exists(PRED_TEMPLATE_FILE):
        return
    with open(PRED_TEMPLATE_FILE, 'r', encoding='utf-8') as f:
        template = f.read()
    html = template.replace('__PREDICTION_DATA__', json.dumps(pred_history))
    with open(PRED_OUTPUT_FILE, 'w', encoding='utf-8') as f:
        f.write(html)
    print('predictions.html を更新しました')

# ============================================================
# 予測スコア計算
# ============================================================
def calc_prediction(history, weights=None):
    if weights is None:
        weights = DEFAULT_WEIGHTS

    all_dates = sorted(set(d for u in history for d in u['days']))
    recent_dates = all_dates[-7:]

    results = []
    for u in history:
        day_results = [u['days'][d][-1] for d in all_dates if d in u['days'] and u['days'][d]]
        if not day_results:
            continue

        recent_results = [u['days'][d][-1] for d in recent_dates if d in u['days'] and u['days'][d]]

        consecutive_loss = 0
        for d in reversed(recent_dates):
            if d in u['days'] and u['days'][d]:
                v = u['days'][d][-1]
                if v < 0:
                    consecutive_loss += 1
                else:
                    break
            else:
                break

        recent_cum_loss = sum(min(v, 0) for v in recent_results)
        avg_daily = sum(day_results) / len(day_results)
        recent_avg = sum(recent_results) / len(recent_results) if recent_results else avg_daily
        wins = sum(1 for v in recent_results if v > 0)
        losses = sum(1 for v in recent_results if v < 0)

        score = (
            consecutive_loss * weights['consecutiveLoss']
            + max(0, -recent_cum_loss) / 1000 * weights['recentCumLoss']
            + max(0, -avg_daily) / 500 * weights['avgDaily']
        )

        results.append({
            'unit': u['unit'],
            'score': round(score * 10) / 10,
            'consecutiveLoss': consecutive_loss,
            'recentAvg': round(recent_avg),
            'recentCumLoss': round(recent_cum_loss),
            'wins': wins,
            'losses': losses,
        })

    return sorted(results, key=lambda x: -x['score'])

# ============================================================
# 当日の実績を過去の予測エントリに記入
# ============================================================
def fill_results(pred_history, history, today_jst):
    for entry in pred_history:
        if entry['predDate'] == today_jst and entry['results'] is None:
            entry['results'] = {}
            for u in history:
                pts = u['days'].get(today_jst)
                if pts:
                    entry['results'][str(u['unit'])] = pts[-1]
            print(f'予測結果記録: {today_jst} ({len(entry["results"])}台)')

# ============================================================
# PDCA: 過去の正解率からスコア重みを調整
# ============================================================
def calc_pdca_weights(pred_history):
    completed = [e for e in pred_history if e.get('results') and len(e['results']) > 0][-14:]
    if len(completed) < 5:
        return DEFAULT_WEIGHTS

    factor_hits = {k: {'hit': 0, 'total': 0} for k in DEFAULT_WEIGHTS}

    for entry in completed:
        hot_preds = entry['predictions'][:3]
        for p in hot_preds:
            result = entry['results'].get(str(p['unit']))
            if result is None:
                continue
            is_hit = result > 0
            for factor in factor_hits:
                if factor in p:
                    factor_hits[factor]['total'] += 1
                    if is_hit:
                        factor_hits[factor]['hit'] += 1

    total_hits = sum(f['hit'] for f in factor_hits.values())
    total_total = sum(f['total'] for f in factor_hits.values())
    base_rate = total_hits / total_total if total_total > 0 else 0.5

    new_weights = dict(DEFAULT_WEIGHTS)
    for factor, stat in factor_hits.items():
        if stat['total'] == 0:
            continue
        hit_rate = stat['hit'] / stat['total']
        multiplier = max(0.5, min(2.0, 0.5 + hit_rate / max(base_rate, 0.1)))
        new_weights[factor] = round(DEFAULT_WEIGHTS[factor] * multiplier * 10) / 10

    return new_weights

# ============================================================
# メイン
# ============================================================
def main():
    fetch_date = datetime.now(JST)
    print(f'[{fetch_date.strftime("%Y/%m/%d %H:%M:%S")}] データ取得開始')

    # 台リストを動的取得
    try:
        active_units = fetch_unit_list()
        print(f'台リスト取得: {len(active_units)}台 [{", ".join(map(str, active_units))}]')
        if os.path.exists(UNITS_FILE):
            with open(UNITS_FILE) as f:
                prev = json.load(f)
        else:
            prev = active_units
        added = [u for u in active_units if u not in prev]
        removed = [u for u in prev if u not in active_units]
        if added:
            print(f'台追加: {", ".join(map(str, added))}')
        if removed:
            print(f'台削除: {", ".join(map(str, removed))}')
        with open(UNITS_FILE, 'w') as f:
            json.dump(active_units, f)
    except Exception as e:
        print(f'警告: 台リスト取得失敗 ({e})。前回リストを使用。')
        if os.path.exists(UNITS_FILE):
            with open(UNITS_FILE) as f:
                active_units = json.load(f)
        else:
            active_units = UNITS_FALLBACK

    new_data = fetch_all(active_units)
    print(f'\n取得完了: {len(new_data)}台')

    existing = []
    if os.path.exists(HISTORY_FILE):
        with open(HISTORY_FILE) as f:
            existing = json.load(f)

    if not new_data:
        print('警告: 新規データなし。既存の history.json でグラフを更新します。')
        if not existing:
            print('エラー: 既存データもありません。終了します。')
            sys.exit(1)
        update_html(existing)
        pred_history = []
        if os.path.exists(PRED_HISTORY_FILE):
            with open(PRED_HISTORY_FILE) as f:
                pred_history = json.load(f)
        update_predictions_html(pred_history)
        print('完了（既存データ使用）')
        return

    history = merge_history(existing, new_data, fetch_date)
    with open(HISTORY_FILE, 'w') as f:
        json.dump(history, f)
    print(f'history.json を更新しました（{len(active_units)}台 × 最大{KEEP_DAYS}日）')

    update_html(history)

    today_jst = get_jst_date_str(fetch_date)
    tomorrow_jst = get_jst_date_str(fetch_date + timedelta(days=1))

    pred_history = []
    if os.path.exists(PRED_HISTORY_FILE):
        with open(PRED_HISTORY_FILE) as f:
            pred_history = json.load(f)

    fill_results(pred_history, history, today_jst)

    weights = calc_pdca_weights(pred_history)
    print(f'PDCA重み: consecutiveLoss={weights["consecutiveLoss"]}, recentCumLoss={weights["recentCumLoss"]}, avgDaily={weights["avgDaily"]}')

    if not any(e['predDate'] == tomorrow_jst for e in pred_history):
        predictions = calc_prediction(history, weights)[:9]
        pred_history.append({
            'predDate': tomorrow_jst,
            'madeAt': today_jst,
            'weights': weights,
            'predictions': predictions,
            'results': None,
        })
        print(f'翌日予測を作成: {tomorrow_jst} (上位{len(predictions)}台)')

    pred_cutoff = get_jst_date_str(fetch_date - timedelta(days=90))
    pruned = [e for e in pred_history if e['predDate'] >= pred_cutoff]

    with open(PRED_HISTORY_FILE, 'w') as f:
        json.dump(pruned, f)
    print('prediction_history.json を更新しました')

    update_predictions_html(pruned)
    print('完了')

if __name__ == '__main__':
    main()
